File: apollo-runtime/src/main/java/com/apollographql/apollo/cache/http/DiskLruHttpCacheStore.java
package com.apollographql.apollo.cache.http;

import java.io.File;
import java.io.IOException;

import javax.annotation.Nonnull;

import okhttp3.internal.cache.DiskLruCache;
import okhttp3.internal.io.FileSystem;
import okio.Sink;
import okio.Source;

public final class DiskLruHttpCacheStore implements HttpCacheStore {
  private static final int VERSION = 99991;
  private static final int ENTRY_HEADERS = 0;
  private static final int ENTRY_BODY = 1;
  private static final int ENTRY_COUNT = 2;

  private final DiskLruCache cache;

  public DiskLruHttpCacheStore(@Nonnull File directory, long maxSize) {
    this.cache = DiskLruCache.create(FileSystem.SYSTEM, directory, VERSION, ENTRY_COUNT, maxSize);
  }

  public DiskLruHttpCacheStore(@Nonnull FileSystem fileSystem, @Nonnull File directory, long maxSize) {
    this.cache = DiskLruCache.create(fileSystem, directory, VERSION, ENTRY_COUNT, maxSize);
  }

  @Override public HttpCacheRecord cacheRecord(@Nonnull String cacheKey) throws IOException {
    final DiskLruCache.Snapshot snapshot = cache.get(cacheKey);
    if (snapshot == null) {
      return null;
    }
    final HttpCacheRecord responseCacheRecord = new HttpCacheRecord() {
      @Nonnull @Override public Source headerSource() {
        return snapshot.getSource(ENTRY_HEADERS);
      }

      @Nonnull @Override public Source bodySource() {
        return snapshot.getSource(ENTRY_BODY);
      }

      @Override public void close() {
        snapshot.close();
      }
    };

    return responseCacheRecord;
  }

  @Override public HttpCacheRecordEditor cacheRecordEditor(@Nonnull String cacheKey) throws IOException {
    final DiskLruCache.Editor editor = cache.edit(cacheKey);
    if (editor == null) {
      return null;
    }

    return new HttpCacheRecordEditor() {
      @Nonnull @Override public Sink headerSink() {
        return editor.newSink(ENTRY_HEADERS);
      }

      @Nonnull @Override public Sink bodySink() {
        return editor.newSink(ENTRY_BODY);
      }

      @Override public void abort() throws IOException {
        editor.abort();
      }

      @Override public void commit() throws IOException {
        editor.commit();
      }
    };
  }

  @Override public void delete() throws IOException {
    cache.delete();
  }

  @Override public void remove(@Nonnull String cacheKey) throws IOException {
    cache.remove(cacheKey);
  }
}

File: deprecated-apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCacheStore.kt
package com.apollographql.apollo3.cache.http

import com.apollographql.apollo3.api.cache.http.HttpCacheRecord
import com.apollographql.apollo3.api.cache.http.HttpCacheRecordEditor
import com.apollographql.apollo3.api.cache.http.HttpCacheStore
import com.apollographql.apollo3.cache.http.internal.DiskLruCache
import com.apollographql.apollo3.cache.http.internal.FileSystem
import okio.Sink
import okio.Source
import java.io.File
import java.io.IOException
import java.util.concurrent.locks.ReadWriteLock
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.read
import kotlin.concurrent.write

class DiskLruHttpCacheStore(private val fileSystem: FileSystem, private val directory: File, private val maxSize: Long) : HttpCacheStore {
  private var cache = createDiskLruCache()
  private val cacheLock = ReentrantReadWriteLock()

  constructor(directory: File, maxSize: Long) : this(FileSystem.SYSTEM, directory, maxSize) {}

  private fun createDiskLruCache(): DiskLruCache {
    return DiskLruCache.create(fileSystem, directory, VERSION, ENTRY_COUNT, maxSize)
  }

  @Throws(IOException::class)
  override fun cacheRecord(cacheKey: String): HttpCacheRecord? {
    val snapshot = cacheLock.read {
      cache[cacheKey]
    }
    return if (snapshot == null) {
      null
    } else object : HttpCacheRecord {
      override fun headerSource(): Source {
        return snapshot.getSource(ENTRY_HEADERS)
      }

      override fun bodySource(): Source {
        return snapshot.getSource(ENTRY_BODY)
      }

      override fun close() {
        snapshot.close()
      }
    }
  }

  @Throws(IOException::class)
  override fun cacheRecordEditor(cacheKey: String): HttpCacheRecordEditor? {
    val editor = cacheLock.read {
      cache.edit(cacheKey)
    }

    return if (editor == null) {
      null
    } else object : HttpCacheRecordEditor {
      override fun headerSink(): Sink {
        return editor.newSink(ENTRY_HEADERS)
      }

      override fun bodySink(): Sink {
        return editor.newSink(ENTRY_BODY)
      }

      @Throws(IOException::class)
      override fun abort() {
        editor.abort()
      }

      @Throws(IOException::class)
      override fun commit() {
        editor.commit()
      }
    }
  }

  @Throws(IOException::class)
  override fun delete() {
    cache = cacheLock.write {
      cache.delete()
      createDiskLruCache()
    }
  }

  @Throws(IOException::class)
  override fun remove(cacheKey: String) {
    // is `read` ok here?
    cacheLock.read {
      cache.remove(cacheKey)
    }
  }

  companion object {
    private const val VERSION = 99991
    private const val ENTRY_HEADERS = 0
    private const val ENTRY_BODY = 1
    private const val ENTRY_COUNT = 2
  }
}